### PR TITLE
[aogh] add aogh target

### PR DESCRIPTION
**Pull request by Chrstm**

### CODE REVIEW

File: examples/apps/aogh/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# OpenThread RTOS AoGH
+
+This example provides a UDP server which receives and handles UDP packets from associated AoGH-based device.
+
+See [Actions on Google - Local execution](https://developers.google.com/actions/smarthome/concepts/local) for more details.
+
+## Build and run the demo
+
+### Prerequisites
+
```
Comment (gjc13) [inline examples/apps/aogh/README.md]: Add instruction on setting up the Google Home device.

File: examples/apps/aogh/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+make -j12
+```
+
+The image will be found at `build/ot_aogh.hex`. Flash the image to NRF52840 dev kit with following command:
+
+```sh
+nrfjprog -f nrf52 --chiperase --program ot_aogh.hex --reset --log
+```
+
+Connect to the serial port of NRF52840 (typically `/dev/ttyACM0`). Log will be directed here.
```
Comment (gjc13) [inline examples/apps/aogh/README.md:22]: Add instruction on playing with AoGH with voice and gui.

File: examples/apps/aogh/aogh.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+#define TURN_ON 0
+#define TURN_OFF 1
+
+#define LED_NUM 4
+
+const uint32_t LED_PIN[4] = {13, 14, 15, 16};
+
+static TaskHandle_t sAoghTask;
+
+// XXX: For quick test only
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: Are these test parameters used?

File: examples/apps/aogh/aogh.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        brightness_ = 0;
+        for (int i = 0; i < LED_NUM; ++i) {
+            nrf_gpio_cfg_output(LED_PIN[i]);
+            nrf_gpio_pin_write(LED_PIN[i], TURN_OFF);
+        }
+    }
+
+    int AoghServer()
+    {
+        printf("AoghServer\n");
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: Clean up debug printfs (useful logs may be retained).

File: examples/apps/aogh/aogh.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        bindAddr.sin6_addr   = IN6ADDR_ANY_INIT;
+        bindAddr.sin6_port   = htons(kAoghPort);
+
+        SuccessOrExit(ret = bind(sockfd, reinterpret_cast<sockaddr *>(&bindAddr), sizeof(bindAddr)));
+
+        while (true)
+        {
+            socklen_t size;
+            printf("recv...\n");
+            VerifyOrExit((size = recvfrom(sockfd, reinterpret_cast<char *>(mBuf), sizeof(mBuf), 0, &peerAddr,
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp:133]: Use `size` and `addrSize` separately.

File: examples/apps/aogh/aogh.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        sendto(aSockFd, &brightness_, sizeof(brightness_), 0, &aPeerAddr, sizeof(aPeerAddr));
+    }
+
+    void GetMacAddress(uint8_t *macAddress)
+    {
+        otExtAddress extAddress;
+        OT_API_CALL(otLinkGetFactoryAssignedIeeeEui64(otrGetInstance(), &extAddress));
+        for (int i = 0, j = 0; i < OT_EXT_ADDRESS_SIZE; ++i)
+        {
+            if (i < 3 || i > 4)
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: Use two `memcpy` instead?

File: examples/apps/aogh/aogh.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            default:
+                printf("[*] Unrecognized packet\n");
+        }
+
+        sendto(aSockFd, &brightness_, sizeof(brightness_), 0, &aPeerAddr, sizeof(aPeerAddr));
+    }
+
+    void GetMacAddress(uint8_t *macAddress)
+    {
+        otExtAddress extAddress;
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp:180]: Rename to EuiToMacAddress?

File: examples/apps/aogh/aogh.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    void LaunchServer()
+    {
+        printf("LaunchServer called\n");
+        AoghServer();
+    }
+
+private:
+    static const uint16_t kAoghPort = 30000;
+    uint8_t brightness_ = 0;
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: Move declaration of members together. Rename to mBrightness.

File: examples/apps/aogh/aogh.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+const otExtendedPanId kTestNetworkXPanId  = {{0xde, 0xad, 0xff, 0xbe, 0xef, 0xff, 0xca, 0xfe}};
+const otMasterKey     kTestNetworkKey     = {
+    {0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff}};
+
+class AoghTask
+{
+public:
+    AoghTask() {}
+
+    void Initialize()
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: `Initialize(void)`
Other functions the same.

File: examples/apps/aogh/aogh.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                break;
+
+            default:
+                printf("[*] Unrecognized packet\n");
+        }
+
+        sendto(aSockFd, &brightness_, sizeof(brightness_), 0, &aPeerAddr, sizeof(aPeerAddr));
+    }
+
+    void GetMacAddress(uint8_t *macAddress)
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: `aMacAddress`, other parameters the same.

File: examples/apps/aogh/aogh.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    static const uint16_t kAoghPort = 30000;
+};
+
+void aoghTask(void *p)
+{
+    (void)p;
+    AoghTask t;
+    t.Initialize();
+    t.LaunchServer();
+    while(true) {
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp]: Format with clang-format

File: examples/apps/aogh/aogh.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                printf("ERROR: unrecognized packet\n");
+        }
+
+        sendto(aSockFd, &mBrightness, sizeof(mBrightness), 0, &aPeerAddr, sizeof(aPeerAddr));
+    }
+
+    void EuiToMacAddress(uint8_t *aMacAddress)
+    {
+        otExtAddress extAddress;
+        OT_API_CALL(otLinkGetFactoryAssignedIeeeEui64(otrGetInstance(), &extAddress));
```
Comment (gjc13) [inline examples/apps/aogh/aogh.cpp:182]: A blank line between declaration and code.
See https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md

**Review by gjc13 — APPROVED** [review]
LGTM

File: examples/apps/aogh/README.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
+
```
Comment (bukepo) [inline examples/apps/aogh/README.md]: remove this empty line?

File: examples/apps/aogh/aogh.cpp [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+#include <openthread/openthread-freertos.h>
+#include <openthread/thread.h>
+
+#include <lwip/sockets.h>
+
+#include <nrfx/hal/nrf_gpio.h>
+
+#include "common/code_utils.hpp"
+#include "utils/thin_function.h"
+#include "../../../third_party/openthread/repo/src/core/common/code_utils.hpp"
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: Is this required?

File: examples/apps/aogh/aogh.cpp [hunk 15]
```diff
@@ -0,0 +1,201 @@
+#include <FreeRTOS.h>
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp:29]: Missing license header.

File: examples/apps/aogh/aogh.cpp [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+
+#include <nrfx/hal/nrf_gpio.h>
+
+#include "common/code_utils.hpp"
+#include "utils/thin_function.h"
+#include "../../../third_party/openthread/repo/src/core/common/code_utils.hpp"
+
+#define MAC_ADDRESS_SIZE 6
+
+#define CMD_OFF 0x00
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: Use `enum` for commands?

File: examples/apps/aogh/aogh.cpp [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+
+#include <lwip/sockets.h>
+
+#include <nrfx/hal/nrf_gpio.h>
+
+#include "common/code_utils.hpp"
+#include "utils/thin_function.h"
+#include "../../../third_party/openthread/repo/src/core/common/code_utils.hpp"
+
+#define MAC_ADDRESS_SIZE 6
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: Use `const size_t` instead?

File: examples/apps/aogh/aogh.cpp [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+
+#define CMD_OFF 0x00
+#define CMD_ON 0x01
+#define CMD_BRIGHTNESS 0x02
+#define FIRST_BYTE_OF_MULTICAST 0x48
+
+#define BRIGHTNESS_MIN 0
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: set proper prefix for these macros?

File: examples/apps/aogh/aogh.cpp [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+#define BRIGHTNESS_MAX 100
+#define TURN_ON 0
+#define TURN_OFF 1
+#define LED_NUM 4
+
+const uint32_t LED_PIN[4] = {13, 14, 15, 16};
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: add static?

File: examples/apps/aogh/aogh.cpp [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+const uint16_t        kTestNetworkPanId   = 0xbeef;
+const uint8_t         kTestNetworkChannel = 14;
+const otExtendedPanId kTestNetworkXPanId  = {{0xde, 0xad, 0xff, 0xbe, 0xef, 0xff, 0xca, 0xfe}};
+const otMasterKey     kTestNetworkKey     = {
+    {0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff}};
+
+class AoghTask
+{
+public:
+    AoghTask() {}
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: missing `void`

File: examples/apps/aogh/aogh.cpp [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+
+    void LaunchServer(void)
+    {
+        AoghServer();
+    }
+
+private:
+    void ConnectToThread(void)
+    {
+        OT_API_CALL(otThreadSetNetworkName(otrGetInstance(), kTestNetworkName));
```
Comment (bukepo) [inline examples/apps/aogh/aogh.cpp]: use `otrLock();` and `otrUnlock();otrTaskNotifyGive();` for invoking a block of OpenThread APIs?

Comment (bukepo) [reply examples/apps/aogh/aogh.cpp]: @gjc13 should we put `otrTaskNotifyGive()` inside of `otrUnlock()`?

Comment (gjc13) [reply examples/apps/aogh/aogh.cpp]: We can 
```
OT_API_CALL(
     otThreadSetNetworkName(...),
     otThreadSetExtendedPanId(...),
.....
);
```

**Review by gjc13 — APPROVED** [review]
LGTM

**Pull request closed without merging**